hookutils: add support for resolving metadata .egg-info files

rokm opened the pull request:
Debian and Ubuntu linux seem to often package python packages with metadata stored in an `.egg-info` file as opposed to `.egg-info` directory or the standard `.dist-info` directory. In such cases, `pkg_resources.get_distribution()` returns an instance of `pkg_resources.EggInfoDistribution` that has `egg_info` field set to `None`, and we need to attempt to resolve the egg-info path
ourselves.

This commit effectively restores a fallback that was in place before #5774, but adds an additional search pattern based on
`project_name` and `version` to supplement the original pattern based on `egg_name()`.

Fixes #6692. The additional search is implemented as a fallback to avoid any interference with the established codepath (which presumably deals only with directories).

rokm committed:
hookutils: add support for resolving metadata .egg-info files

Debian and Ubuntu linux seem to often package python packages
with metadata stored in an .egg-info file as opposed to .egg-info
directory or the standard .dist-info directory. In such cases,
pkg_resources.get_distribution() returns an instance of
pkg_resources.EggInfoDistribution that has egg_info field set
to None, and we need to attempt to resolve the egg-info path
ourselves.

This commit effectively restores a fallback that was in place
before #5774, but adds an additional search pattern based on
project_name and version to supplement the original pattern
based on egg_name().

bwoodsend commented:
If you copy/paste one of these distributions with a .egg-info file somewhere into the `tests` folder and `git add -f` it in (or maybe write an exclusion in the .gitignore), is it possible to get a test in there? This is exactly the kind of niche thing I'd end up destroying the next time I try cleaning something else up.

rokm commented:
Fair enough, I'll see if I can come up with a reasonable test for this.

On an unrelated note, what's with the "too many files open" error? I've seen it pop up couple of times now (and I think it was always 3.9 on Windows) - are we leaking handles somewhere?

bwoodsend commented:
Ohh, that reminds me - when I finally got round to giving our NumPy hook to NumPy, their test suite had some kind of file handle leak detecter which said that we were. I _think_ I had to tell it to exclude the PyInstaller test I submitted with it. Let me check...

bwoodsend commented:
Yeah, I did as well: 
https://github.com/numpy/numpy/pull/20745/files#diff-a2750c3ab420161baf467e763884086ad8070783d86dc67376763f24f85c8920R9-R10

It was a warning and NumPy were using some *warnings are errors* flag. I'll see if I can find what's doing the leak detection.

bwoodsend commented:
Turns out it's just Python doing it by itself. The reason we don't see it is because `ResourceWarning` is ignored by default. Adding `-W 'error::ResourceWarning'` to either Python or `pytest` should tell us where the leaks are.

bwoodsend commented:
Do `io.BytesIO()` instances need closing? I thought that they were just a mock file-like wrapper around a string but this seems to say otherwise?
```python
/e/n/pyinstaller (pyup/scheduled-update-2022-03-06)> pytest -W 'error::ResourceWarning' -k basic tests/functional/test_basic.py::test_run_from_path_environ
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/brenainn/.pyenv/versions/3.9.9/bin/python3.9
cachedir: .pytest_cache
rootdir: /e/notebooks/pyinstaller, configfile: setup.cfg
plugins: rerunfailures-10.2, stress-1.0.1, drop-dup-tests-0.3.0, cookies-0.6.1, flaky-3.7.0, order-1.0.1, forked-1.4.0, xdist-2.5.0, timeout-2.1.0, cov-3.0.0
timeout: 600.0s
timeout method: signal
timeout func_only: False
collected 2 items                                                                                                                                                                                                                            

tests/functional/test_basic.py::test_run_from_path_environ[onedir] PASSED                                                                                                                                                              [ 50%]
tests/functional/test_basic.py::test_run_from_path_environ[onefile] PASSED                                                                                                                                                             [100%]

============================================================================================================== warnings summary ==============================================================================================================
PyInstaller/lib/modulegraph/modulegraph.py:20
  /e/notebooks/pyinstaller/PyInstaller/lib/modulegraph/modulegraph.py:20: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/functional/test_basic.py::test_run_from_path_environ[onedir]
tests/functional/test_basic.py::test_run_from_path_environ[onefile]
  /home/brenainn/.pyenv/versions/3.9.9/lib/python3.9/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <_io.FileIO [closed]>
  
  Traceback (most recent call last):
    File "/e/notebooks/pyinstaller/PyInstaller/archive/writers.py", line 377, in add
      fh = io.BytesIO(data)
  ResourceWarning: unclosed file <_io.BufferedReader name='/e/notebooks/pyinstaller/PyInstaller/loader/pyimod04_ctypes.pyc'>
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================================================================================================= 2 passed, 3 warnings in 19.76s =======================================================================================================
```

bwoodsend commented:
Ohh no. It's because `fh = io.BytesIO(data)` clobbers the `fh = open(pathnm, 'rb')` reference from a few lines up. The original `fh` in not closed.

rokm committed:
tests: add a test for single-file metadata collection

rokm committed:
building: reset pkg_resources working set at start of the build

Required for pkg_resources to register changes in sys.path during
long-running sessions (e.g., multiple builds during our CI run).

bwoodsend commented:
I'm fixing the file handles leaks.

rokm committed:
tests: fix copy of test's .egg-info file in Docker container

Due to the blanket ignore rule for *.egg-info patterns in
.dockerignore file, the  my_test_package-1.0.egg-info file used
by the test_misc::test_single_file_metadata does not end up
copied into the test container, causing the corresponding test
to fail.
